From c30228f69b0aab8cdcca55adad50a817df8121da Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:27:07 -0400
Subject: [tools] perf/traceevent: If %s is a pointer, check printk formats

Message-id: <1408462094-14194-259-git-send-email-jolsa@redhat.com>
Patchwork-id: 88312
O-Subject: [PATCH RHEL7.1 BZ1131394 258/325] tools lib traceevent: If %s is a pointer, check printk formats
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 0970b5f438261216afcd0ccaa2fcfffc83df7ca2
Author: Steven Rostedt (Red Hat) <rostedt@goodmis.org>
Date: Fri Nov 1 17:53:55 2013 -0400

description
===========
If the format string of TP_printk() contains a %s, and the argument is
not a string, check if the argument is a pointer that might match the
printk_formats that were stored.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/event-parse.c b/tools/lib/traceevent/event-parse.c
index 856b791..013c8d3 100644
--- a/tools/lib/traceevent/event-parse.c
+++ b/tools/lib/traceevent/event-parse.c
@@ -3505,6 +3505,7 @@ static void print_str_arg(struct trace_seq *s, void *data, int size,
  struct pevent *pevent = event->pevent;
  struct print_flag_sym *flag;
  struct format_field *field;
+ struct printk_map *printk;
  unsigned long long val, fval;
  unsigned long addr;
  char *str;
@@ -3540,7 +3541,12 @@ static void print_str_arg(struct trace_seq *s, void *data, int size,
   if (!(field->flags & FIELD_IS_ARRAY) &&
       field->size == pevent->long_size) {
    addr = *(unsigned long *)(data + field->offset);
-   trace_seq_printf(s, "%lx", addr);
+   /* Check if it matches a print format */
+   printk = find_printk(pevent, addr);
+   if (printk)
+    trace_seq_puts(s, printk->printk);
+   else
+    trace_seq_printf(s, "%lx", addr);
    break;
   }
   str = malloc(len + 1);
-- 
1.7.1